client-v2/src/main/java/com/clickhouse/client/api/data_formats/internal/BinaryStreamReader.java

```diff
@@ -543,4 +673,42 @@ private static int readByteOrEOF(InputStream input) throws IOException {
         }
         return b;
     }
+
+    public interface ByteBufferAllocator {
+        byte[] allocate(int size);
+    }
+
+    /**
+     * Byte allocator that creates a new byte array for each request.
+     */
+    public static class DefaultByteBufferAllocator implements ByteBufferAllocator {
+        @Override
+        public byte[] allocate(int size) {
+            return new byte[size];
+        }
+    }
+
+    /**
+     * Byte allocator that caches preallocated byte arrays for small sizes.
+     */
+    public static class CachingByteBufferAllocator implements ByteBufferAllocator {
+
+        private static final int MAX_PREALLOCATED_SIZE = 32;
+        private final byte[][] preallocated = new byte[MAX_PREALLOCATED_SIZE + 1][];
+
+        public CachingByteBufferAllocator() {
+           for (int i = 0; i < preallocated.length; i++) {
+               preallocated[i] = new byte[i];
+           }
+        }
+
+        @Override
+        public byte[] allocate(int size) {
+            if (size < preallocated.length) {
+                return preallocated[size];
+            }
+
+            return new byte[size];

```

Comment:
In what cases can the size bigger larger than `MAX_PREALLOCATED_SIZE` + 1?
if it meets the case, it always allocates byte[size]. Do we want to allocate and cache the array for future usage?

Reply:
looking on `return readNBytesLE(input, bufferAllocator.allocate(len), 0, len);` this the only place that we can hit 
`return new byte[size];` Maybe we also need to reuse it? I think if we have a table with a lot of strings bigger than 32 we actually will allocate every time and since we use it between rows I think the change that we hit same size allocation is high.

Reply:
This implementation is to solve only small buffer allocations for numbers. 
As for Strings, we may be need another approach. If string is small - we can cache them, if big - we would need to represent them with another class (`String` copies source buffer) but it will be confusing.

Reply:
This what i had in my mind we can do it later